extension/httpfs/httpfs_client.cpp

```diff
@@ -2,146 +2,499 @@
 		}
+
+		const char* data = request_info->body.c_str();
+		info.content_handler(const_data_ptr_cast(data), bytes_received);
+		return TransformResponseCurl(std::move(request_info), std::move(header_collector), res);
 	}
+
 	unique_ptr<HTTPResponse> Put(PutRequestInfo &info) override {
 		if (state) {
 			state->put_count++;
 			state->total_bytes_sent += info.buffer_in_len;
 		}
-		auto headers = TransformHeaders(info.headers, info.params);
-		return TransformResult(client->Put(info.path, headers, const_char_ptr_cast(info.buffer_in), info.buffer_in_len,
-		                                   info.content_type));
+
+		auto curl_headers = TransformHeadersCurl(info.headers);
+		auto request_info = make_uniq<RequestInfo>();
+		// Add content type header from info
+		curl_headers.Add("Content-Type: " + info.content_type);
+		// transform parameters
+		request_info->url = info.url;
+		if (!info.params.extra_headers.empty()) {
+			auto curl_params = TransformParamsCurl(info.params);
+			request_info->url += "?" + curl_params;
+		}
+
+		auto header_collector = make_uniq<HeaderCollector>();
+		CURLcode res;
+		{
+			curl_easy_setopt(*curl, CURLOPT_URL, request_info->url.c_str());
+
+			// Perform PUT
+			curl_easy_setopt(*curl, CURLOPT_CUSTOMREQUEST, "PUT");
+
+			// Include PUT body
+			curl_easy_setopt(*curl, CURLOPT_POSTFIELDS, const_char_ptr_cast(info.buffer_in));
+			curl_easy_setopt(*curl, CURLOPT_POSTFIELDSIZE, info.buffer_in_len);
+
+			// Follow redirects if needed
+			curl_easy_setopt(*curl, CURLOPT_FOLLOWLOCATION, 1L);
+
+			// Capture response body
+			curl_easy_setopt(*curl, CURLOPT_WRITEFUNCTION, RequestWriteCallback);
+			curl_easy_setopt(*curl, CURLOPT_WRITEDATA, &request_info->body);
+
+			// Capture response headers
+			curl_easy_setopt(*curl, CURLOPT_HEADERFUNCTION, RequestHeaderCallback);
+			curl_easy_setopt(*curl, CURLOPT_HEADERDATA, header_collector.get());
+
+			// Apply headers
+			curl_easy_setopt(*curl, CURLOPT_HTTPHEADER, curl_headers ? curl_headers.headers : nullptr);
+
+			// Execute the request
+			res = curl->Execute();
+		}
+
+		curl_easy_getinfo(*curl, CURLINFO_RESPONSE_CODE, &request_info->response_code);
+
+		return TransformResponseCurl(std::move(request_info), std::move(header_collector), res);
 	}
 
 	unique_ptr<HTTPResponse> Head(HeadRequestInfo &info) override {
 		if (state) {
 			state->head_count++;
 		}
-		auto headers = TransformHeaders(info.headers, info.params);
-		return TransformResult(client->Head(info.path, headers));
+
+		auto curl_headers = TransformHeadersCurl(info.headers);
+		auto request_info = make_uniq<RequestInfo>();
+		request_info->url = info.url;
+		// transform parameters
+		if (!info.params.extra_headers.empty()) {
+			auto curl_params = TransformParamsCurl(info.params);
+			request_info->url += "?" + curl_params;
+		}
+
+		CURLcode res;
+		auto header_collector = make_uniq<HeaderCollector>();
+		{
+			// Set URL
+			curl_easy_setopt(*curl, CURLOPT_URL, request_info->url.c_str());
```

Comment:
There's now quite a bit of repitition in these curl configurations between the different request methods. Could we perhaps move the shared config between them in a function?